### ISSUE TITLE
Dynamic adding of mappings

### ISSUE THREAD

**Issue by siberianguy** [AutoMapper/AutoMapper#2317]
For years I've been hesitant about using AutoMapper because of its indirect nature: somewhere in one big file there're all the mappings and it's not clear what's going on when you read the method that uses Map. Introducing Profiles improved this a lot as now we can at least group mappings into groups.

Recently I've been looking into Jimmy's [ContosoUniversityCore project](https://github.com/jbogard/ContosoUniversityCore) and realised mappings are not getting reused as for every action there's one command/query and one handler. So it makes no sense to store mappings together in a Profile file common for the whole feature. Mapping belongs to the file with command/query and handler. Jimmy seems to agree with this: https://github.com/jbogard/ContosoUniversityCore/issues/19

But what is the best way to implement it? Of course we could add a Profile class right next to Command/Query and Handler but it seems to be an overkill for simple scenarios:
```
public class MappingProfile : Profile
    {
        public MappingProfile()
        {
            CreateMap<Course, Index.Result.Course>();
        }
    }
```
So 6 decorative lines just to add a one-line mapper.

What might be a better solution to this specific problem is adding inline mappers.
```
       public class Handler : IRequestHandler<Command>
        {
            private readonly SchoolContext _db;

            public Handler(SchoolContext db)
            {
                _db = db;
            }

            public void Handle(Command message)
            {
                **Mapper.CreateMap<Course, Index.Result.Course>();** //the only new line
                var course = Mapper.Map<Command, Course>(message);
                course.Id = message.Number;

                _db.Courses.Add(course);
            }
        }
```
In the example above I just took [this handler](https://github.com/jbogard/ContosoUniversityCore/blob/master/src/ContosoUniversityCore/Features/Course/Create.cs) and added one line that CreatesMap.

This approach (specific syntax is debatable) would solve the indirection that bothered me about AutoMapper for years. Here's the mapping definition, right where it gets used. It's as clear as writing manual mappings but you don't have to write repetitive code: the best from both worlds.

I assume different kinds of inline mappings were discussed multiple times during AutoMapper development and there should be good reasons why it hasn't been implemented (or am I missing something?). Implementations thoughts: CreateMap should check if this mapping has already added and don't add it again, so there should not be a huge overhead; Mapper.CreateMap doesn't have to be static: we could inject mapper in some form.

**Comment by lbargaoanu** [comment]
This won't work in the general case. You can do smth like that today with AddConditionalObjectMapper or CreateMissingTypeMaps. But when/if you get to more complex cases, you're on your own.

**Comment by siberianguy** [comment]
@lbargaoanu But why? Seems like a reasonable use case for me. Are there any fundamental limitations? Is it realistic to refactor existing codebase to natively support this case?

**Comment by lbargaoanu** [comment]
This has been discussed many times before and I kind of lost interest :) @jbogard might have more to say on the matter.

**Comment by siberianguy** [comment]
I thought about it a little more and even the separate method for mapping registration makes no sense here. It could be just:
 var course = Mapper.MapInline<Command, Course>(message)
   .ForMember(...);

**Comment by jbogard** [comment]
So just as a general question, what should happen in the case where the mapping is invalid, or wouldn't pass configuration validation? Throw an exception?

**Comment by jbogard** [comment]
And if you're asking for "why doesn't AutoMapper support this", well, it kinda does as was pointed out above. But then you coooooompletely lose the ability to do one-line testing of mapping configuration `Mapper.AssertConfigurationIsValid`. Half the reason AutoMapper exists is to make it easy to test your configuration - that you've not misspelled anything or accidentally left a member there, or a dozen other things that could happen.

This might be an interesting use case if by default we validated the mappings the first time it ran (and provided a special overload, like InlineMap etc)

**Comment by TylerCarlson1** [comment]
I would suggest `AddConditionalObjectMapper()` and make a convention that follows the logic of the handlers.

For @jbogard's example
Like Namespace has to be "ContosoUniversityCore.Features." + EFClass namespace and inherit from IRequest. will make a default mapping.  If you need something more specific you create a profile class to show the more complicated logic in that class.  Mostly it should be simple and the convention should suffice, but if you see a profile that means you had to do something more specific in the mapping, or an indication someone didn't follow the convention you want to adhere to.

The Convention will be in a separate file independent of the mappings, but it's a general rule that you should follow, and should work for most if not all scenarios.  Just like how EntityFramework has some default rules.  Like Id or ClassNameID will default it to an Auto Incriminating Primary Key, and if you want something different you will have to manually configure it with configuration code on the DataContext, or with Attributes.

If you want everything to follow Mediator pattern in ContosoUniversityCore example then you would want all names to match and be correct everywhere.  With a convention then if it didn't work exactly as you specified it should, then you or someone else did something wrong, and they should follow the pattern to fix it, or make a profile and be punished with ugly code for not following the rules.

It's a trade off, but saves you even more lines of code.

**Comment by jbogard** [comment]
I think this is a bit different than conventions, though, you're also asking to be able to configure the map inline.

Wouldn't this be a bit distracting, though? What would you do if you had more configuration? Opt for profiles I guess?

**Comment by siberianguy** [comment]
The issue was closed, so I assume there's no interest in this feature?

**Comment by lbargaoanu** [comment]
I closed it because it seemed stale :) If you're interested in writing the code you should first convince @jbogard.

**Comment by jbogard** [comment]
Can we open a new issue to discuss mechanics? If our default behavior is to do configuration validation for new maps created, maybe it's not absolutely terrible :)

**Comment by siberianguy** [comment]
I'm not aware of the workflow you guys are used to here, so I hope @lbargaoanu might be kind enough to open a technical issue.

> maybe it's <s>not absolutely terrible</s> amazing :)

Correct me if I'm wrong but this inline approach might make things simpler for more complex scenarios, i. e. when you need to use DataContext for a mapping (like when you have ItemId but you need to map it into Item). I'm looking at the codebase that I'm refactoring and I see it's done in an ugly way: DataContext is resolved at StartUp and is used for those kinds of mappings, though perfectly a PerThread DataContext should be used. I don't have particular example but I feel like this approach will give wrong results in some scenarios. I don't know how it's done correctly in AutoMapper but I assume it's something cumbersome. In case of mappings in a Handler we can get inlined access to current DataContext.

**Comment by jbogard** [comment]
There's already a built-in way to handle contextual data in AutoMapper, and I'd be suuuuuuuuper worried about capturing a DataContext in a closure and having that DataContext stay around...forever? Profile instances aren't disposed, I don't think you want them having anything but static/singleton dependencies. For the built-in ways, see:

https://github.com/AutoMapper/AutoMapper/wiki/Custom-value-resolvers#passing-in-key-value-to-mapper

https://github.com/AutoMapper/AutoMapper/wiki/Queryable-Extensions#parameterization

And of course, using custom value resolvers and injecting the DbContext directly into it.

Anyone can open a GH issue for a feature request, FYI.

**Comment by siberianguy** [comment]
It seems like we have two different implementations in our heads. I was thinking (at least when writing the previous message) about a completely inline implementation where mapping is not stored. I guess there's overhead to that but how bad is it? Is it bad enough to not be able to write code which is very similar to manual mappings (i. e. you get access to local variables)?

I would love to create a feature request but I'm confused why we need a new thread for it. I feel like we have all we need here: code samples, implementations in alternative tools, etc.

**Comment by jbogard** [comment]
Compile the expression tree every single time? That's pretty bad, expression tree compilation is quite CPU intensive. The map would still need to be built and compiled every time. It would need to validate the map too the first time.

**Comment by jbogard** [comment]
So what I'm looking at here is adding an option to the configuration options, something like:

```
var dest = Mapper.Map<Source, Dest>(src, opt => opt.Configure(cfg => cfg.ForMember(...)));
```

**Comment by siberianguy** [comment]
The syntax looks awesome. We could also have a `Mapper.Map<Dest>(...)` overload for scenarios where `Source` is exactly the type of `src` (that's the case for the CQRS scenario we started discussion with).

Dynamic/static mappings seems like an unnatural separation to me, especially if we introduce this feature. All the mappings are dynamic, they're just created at different times (i. e. startup, method call, etc.). AutoMapper.Initialize always felt too singletonish to me.

With this feature we also might improve the way how dependencies are injected. Currently at usage time you have to push a variable with the same name as the variable introduced during mapping registration. I hope we can agree it's not the most beautiful part of AutoMapper design. With local dynamic mappings we have exactly the variable we want to inject (i. e. _dataContext variable of a Command), so we can do it in a strongly-typed manner. The only question to me is how to tell AutoMapper not to use the variable itself (as we don't want this specific _dataContext to be reused for different queries) but just keep the name of it and retrieve it from the local context. Honestly, not even sure if it's possible in C#.

**Comment by lbargaoanu** [comment]
@jbogard CreateMissingTypeMaps true by default? I think it will be pretty bad :)

**Comment by TylerCarlson1** [comment]
I agree with @lbargaoanu CreateMissingTypeMaps true by default is a bad idea, because it causes so many issues in use cases where you wouldn't think it would.  Like proxy types in EF being mapped instead of the maps you set up for it.

**Comment by siberianguy** [comment]
I feel like we're talking about two different "dynamic" things: one is to autogenerate mapper if it was not created (agree, bad idea, at least by default), another one is registering mappers statically before usage (seems outdated and singletonish to me).

**Comment by TylerCarlson1** [comment]
I'm opposed to the AutoGenerate idea @jbogard said.  I'm also opposed to the Dynamic Registration you are talking about, but CreateMissingTypeMaps = true by default is just a horrible idea all around and doesn't solve any problems.

Also another possible issue just thought about is what if you have single Registered Map that require maps for objects that are also single Registered somewhere else in code.  How do you guarantee that they are getting generated first?  What about circular references or MaxDepth and all those?  You kinda need to know what all the mappings configurations are before you can generate an execute path for it.  Otherwise you probably get an exception trying to create the execution path or running it for the first time.

This probably won't happen in you as you could account for it.  But if others do it this way they might abuse it and come into these scenarios and wonder why this new functionality breaks AutoMapper so much.

This still is just a guess, but if it is then we would need to come to consensus on how to address these issues.

**Comment by jbogard** [comment]
There’s really no difference between those two options.

**Comment by jbogard** [comment]
So I gave it a go on a local branch. Many, many failing tests. Mostly around maps getting created for derived/proxy types then not inheriting any base configuration. Pretty hard to figure out, I don't know how people use `CreateMissingTypeMaps` in the wild. There are a ton of gotchas.

Or maybe people add explicit maps when things go off the rails? But in most cases it just blows up.

🤷‍♂️

**Comment by siberianguy** [comment]
This is sad. I guess will have to switch to one of the alternative mappers, most of them support it in one or another way.

**Comment by jbogard** [comment]
Putting this in breaks dozens of tests.

Maybe with an explicit overload instead or differently named method?

We used to have Mapper.DynamicMap but I wound up removing it because there
were soooooo many edge cases.

**Comment by sellomkantjwa** [comment]
>I don't know how people use CreateMissingTypeMaps in the wild. There are a ton of gotchas. Or maybe people add explicit maps when things go off the rails? But in most cases it just blows up.

@jbogard We do exactly this. We have 2 namespaces that have hundreds of classes that have the same names and the same members - basically, besides the fact that they are from different namespaces, the types are identical. I assumed the choice was between writing hundreds of mappings explicitly or setting `CreateMissingTypeMaps`. Is there a better way?

**Comment by TylerCarlson1** [comment]
@sellomkantjwa 
```
Mapper.Initialize(cfg => 
{
   cfg.AddConditionalObjectMapper().Where((s, d) => s.Namespace != d.Namespace && s.Name == d.Name);
});
```
This creates maps only for types that have matching names but in different namespaces.  If you want to go more specific you can specify source and destination namespaces and check so that the mappings either go one way or the other, or only types in these namespaces can be mapped when match.

This way you don't have any gotchas like CreateMissingTypeMaps

**Comment by jbogard** [comment]
@sellomkantjwa I've created a branch where I basically turn on `CreateMissingTypeMaps` by default. One other thing I did though is to turn on configuration validation by default.

One critical feature of AutoMapper is the configuration validation. I wouldn't really use it otherwise, you'd just get silent missing members.

It might be interesting to see what inline configuration might look like, so I'm running some reports on large projects we have to see what it might look like to inline map these things.

There are two feature requests here:

 - Allow inline mapping configuration (during Mapper.Map)
 - Allow "better" automatic type map creation

Configuration validation right now kinda assumes you've got the entire type map and its children created, so that's blocking a bit.
